File: src/io/innofang/Builder/Audi.java
package io.innofang.Builder;

/**
 * Created by Inno Fang on 2017/3/11.
 */
public class Audi {

    private String color;
    private String licensePlate;
    private String brand;

    public static class Builder {

        private String color;
        private String licensePlate;
        private String brand;

        public Builder setColor(String color) {
            this.color = color;
            return this;
        }

        public Builder setLicensePlate(String licensePlate) {
            this.licensePlate = licensePlate;
            return this;
        }

        public Builder setBrand(String brand) {
            this.brand = brand;
            return this;
        }

        public Audi build(){
            return new Audi(this);
        }
    }

    private Audi(Builder builder) {
        color = builder.color;
        licensePlate = builder.licensePlate;
        brand = builder.brand;
    }

    @Override
    public String toString() {
        return "Car{" +
                "color='" + color + '\'' +
                ", licensePlate='" + licensePlate + '\'' +
                ", brand='" + brand + '\'' +
                '}';
    }
}

File: src/io/innofang/builder/example/kotlin/Audi.kt
package io.innofang.builder.example.kotlin

/**
 * Created by Inno Fang on 2017/8/31.
 */
data class Audi
private constructor(var color: String, var licensePlate: String, var brand: String) {

    private constructor(builder: Builder) : this(
            builder.color,
            builder.licensePlate,
            builder.brand
    )

    class Builder {

        lateinit var color: String
        lateinit var licensePlate: String
        lateinit var brand: String

        fun color(init: Builder.() -> String) = apply { color = init() }
        fun licensePlate(init: Builder.() -> String) = apply { licensePlate = init() }
        fun brand(init: Builder.() -> String) = apply { brand = init() }
    }

    companion object {
        fun build(init: Builder.() -> Unit) = Audi(Builder().apply(init))
    }
}